src/core/thread/link_quality.cpp

```diff
@@ -0,0 +1,274 @@
+/*
+ *  Copyright (c) 2016, Nest Labs, Inc.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements link quality information processing and storage.
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <openthread-types.h>
+#include <common/code_utils.hpp>
+#include <thread/link_quality.hpp>
+
+namespace Thread {
+
+// This array gives the decimal point digits representing 0/8, 1/8, ..., 7/8 (it does not include the '.').
+static const char *const kLinkQualityDecimalDigitsString[8] =
+{
+    // 0/8, 1/8,   2/8,   3/8,   4/8,   5/8,   6/8,   7/8
+    "000", "125", "250", "375", "500", "625", "750", "875"
+};
+
+LinkQualityInfo LinkQualityInfo::sNoiseFloorAverage;
+
+const char LinkQualityInfo::kUnknownRssString[] = "Unknown RSS";
+
+//-------------------------------------------------------------------------------
+
+LinkQualityInfo::LinkQualityInfo(void)
+{
+    Clear();
+}
+
+void LinkQualityInfo::Clear(void)
+{
+    mRssAverage = 0;
+    mCount = 0;
+    mLinkQuality = 0;
+}
+
+ThreadError LinkQualityInfo::AddRss(int8_t anRss)
+{
+    ThreadError error = kThreadError_None;
+    uint16_t    newValue;
+    uint16_t    oldAverage;
+
+    VerifyOrExit(anRss < 0, error = kThreadError_InvalidArgs);
```

Comment:
Good question. 

Practically I don't expect any 15.4 radio to give us a positive RSS. The main reason I restricted the positives was that I wanted to the average to fit in 10 bits (with the goal to keep the total link info data per neighbor to fit in 16 bit).

How about if we see a positive RSS we max/cut it to zero (and not consider it invalid)? Does this make sense?